README
NAME

    Class::Singleton - Implementation of a "Singleton" class

SYNOPSIS

    use Class::Singleton;

    my $one = Class::Singleton->instance();   # returns a new instance
    my $two = Class::Singleton->instance();   # returns same instance

DESCRIPTION

    This is the Class::Singleton module. A Singleton describes an object class
    that can have only one instance in any system. An example of a Singleton
    might be a print spooler or system registry. This module implements a
    Singleton class from which other classes can be derived. By itself, the
    Class::Singleton module does very little other than manage the instantiation
    of a single object. In deriving a class from Class::Singleton, your module
    will inherit the Singleton instantiation method and can implement whatever
    specific functionality is required.

INSTALLATION

    See the INSTALL file.

COPYRIGHT

    Copyright (C) 1998 Canon Research Centre Europe Ltd.
    Copyright (C) 1998-2008 Andy Wardley. All rights reserved.
    Copyright (C) 2014, 2020 Steve Hay. All rights reserved.

LICENCE

    This distribution is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, i.e. under the terms of either the GNU
    General Public License or the Artistic License, as specified in the LICENCE
    file.


Changes
Revision history for Perl distribution Class-Singleton

1.7 Development

    - TODO

1.6 2020-12-02

    - Fixed confusing Changes entry about Perl's licensing terms.
      [<KENTNL@cpan.org>, CPAN RT#132843]

    - Added optional Changes testing (skipped unless AUTHOR_TESTING).

    - Reformatted Changes file as per CPAN::Changes::Spec.

    - Added optional POD coverage testing (skipped unless AUTHOR_TESTING).

    - Added optional Perl::Critic testing (skipped unless AUTHOR_TESTING).

    - Made code Perl::Critic clean.

    - Added optional POD testing (skipped unless AUTHOR_TESTING).

    - Included GitHub repository URLs in metadata now that source code has been
      uploaded to GitHub (as of version 1.5).

    - Included META.json file in addition to META.yml.

    - Set minimum required ExtUtils::MakeMaker version to 6.64 to ensure that
      all parameters used are supported, to save jumping through hoops to
      support earlier versions.  (This should not be a problem since
      ExtUtils::MakeMaker 6.64 is easily installed into Perl 5.8.1 and above,
      that being the whole point of the new choice of minimum supported Perl
      version.)

    - Set minimum required Perl version to 5.8.1.  This is in line with the
      minimum requirement of the "Perl Toolchain".

    - Corrected typo in a comment.  [<dsteinbrunner@pobox.com>, CPAN RT#86336]

1.5 2014-11-07

    - Work around global destruction order issue.  [Craig Manley
      <cmanley@cpan.org>, CPAN RT#23568/68526]

1.4 2007-09-28

    - Added the has_instance() method to return an existing instance without
      creating a new one.

    - General cleanup of code, documentation and tests.

    - Changed licence from Perl Artistic to the same terms as Perl itself (i.e.
      Artistic 1.0 or GPL v1+).

1.03 1999-01-19

    - Incorporated patches from Andreas Koenig to inline calculation of
      $instance variable.  This results in a speedup of around 35%.

    - Added _new_instance() constructor which is called the first time
      _instance() is called.  This can be overloaded in derived classes to
      provide more specific object initialisation.

    - Updated documentation accordingly.

1.02 1998-04-16

    - Fixed minor typos and corrected example in documentation.

1.01 1998-02-10

    - Minor documentation changes

1.00 1998-02-10

    - Initial revision


test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

BEGIN { $| = 1; print "1..20\n"; }
END {print "not ok 1\n" unless $loaded;}
use Class::Singleton;
$loaded = 1;
print "ok 1\n";

######################### End of black magic.

# Insert your test code below (better if it prints "ok 13"
# (correspondingly "not ok 13") depending on the success of chunk 13
# of the test code):

# turn warnings on
$^W = 1;


#========================================================================
#
# define 'MySingleton', a class derived from Class::Singleton 
#
#========================================================================

package DerivedSingleton;

use vars qw(@ISA);
@ISA = qw(Class::Singleton);



#========================================================================
#
# define 'MyOtherSingleton', a class derived from MySingleton 
#
#========================================================================

package AnotherSingleton;

use vars qw(@ISA);
@ISA = qw(DerivedSingleton);



#========================================================================
#
# define 'UniqueSingleton', where we adapt instance() to only allow
# one instance of all and any derived classes.  We do this by forcing 
# all derived classes to call the Class::Singleton instance() method
# masqueraded as UniqueSingleton->instance() (no matter what the derived
# class name might be).  The reference returned is then re-blessed into
# the current derived class.
#
#========================================================================

package UniqueSingleton;

use vars qw(@ISA);
@ISA = qw(Class::Singleton);

sub instance {
    my $self  = shift;
    my $class = ref($self) || $self;

    # create a temporary $self instance blessed into the current
    # UniqueSingleton class so that all derived classes look like a 
    # 'UniqueSingleton' when they call Class::Singleton::instance())
    $self = bless {};

    # call Classs::Singleton->instance() while masquerading...
    $self = $self->SUPER::instance();

    # ...and then bless returned instance into the required class
    bless $self, $class;
}


#========================================================================
#
# create two classes derived from UniqueSingleton which should then be
# mutually exclusive
#
#========================================================================

package UniqueSingletonOne;

use vars qw(@ISA);
@ISA = qw(UniqueSingleton);

sub one { my $self = shift; print "$self->one(", join(', ', @_), ")\n"; };

#------------------------------------------------------------------------

package UniqueSingletonTwo;

use vars qw(@ISA);
@ISA = qw(UniqueSingleton);

sub two { my $self = shift; print "$self->two(", join(', ', @_), ")\n"; };


#========================================================================
#
# We should be able to create one and only once instance of each of
# Class::Singleton, MySingleton and MyOtherSingleton.
#
#========================================================================

package main;

# call Class::Singleton->instance() twice and expect to get the same 
# reference returned on both occasions.

my $s1 = Class::Singleton->instance();

print "Class::Singleton instance 1: ",
    defined($s1) ? ok($s1) : not_ok('<undef>');

my $s2 = Class::Singleton->instance();

print "Class::Singleton instance 2: ",
    (defined($s2) ? ok($s2) : not_ok('<undef>'));

print $s1 == $s2 
    ? ok('Class::Singleton instances are identical') 
    : not_ok('Class::Singleton instances are unique');


# call MySingleton->instance() twice and expect to get the same 
# reference returned on both occasions.

my $s3 = DerivedSingleton->instance();

print "DerivedSingleton instance 1: ", 
    defined($s3) ? ok($s3) : not_ok('<undef>');

my $s4 = DerivedSingleton->instance();

print "DerivedSingleton instance 2: ", 
    defined($s4) ? ok($s4) : not_ok('<undef>');

print $s3 == $s4 
    ? ok("DerivedSingleton instances are identical")
    : not_ok("DerivedSingleton instances are unique");


# call MyOtherSingleton->instance() twice and expect to get the same 
# reference returned on both occasions.

my $s5 = AnotherSingleton->instance();

print "AnotherSingleton instance 1: ",
    defined($s5) ? ok($s5) : not_ok('<undef>');

my $s6 = AnotherSingleton->instance();

print "AnotherSingleton instance 2: ",
    defined($s6) ? ok($s6) : not_ok('<undef>');

print $s5 == $s6 
    ? ok("AnotherSingleton instances are identical")
    : not_ok("AnotherSingleton instances are unique");


#========================================================================
#
# having checked that each instance of the same class is the same, we now
# check that the instances of the separate classes are actually different 
# from each other 
#
#========================================================================

print $s1 != $s3 
    ? ok("Class::Singleton and DerviedSingleton are different") 
    : not_ok("Class::Singleton and DerivedSingleton are identical");
print $s1 != $s5 
    ? ok("Class::Singleton and AnotherSingleton are different") 
    : not_ok("Class::Singleton and AnotherSingleton are identical");
print $s3 != $s5 
    ? ok("DerivedSingleton and AnotherSingleton are different") 
    : not_ok("DerivedSingleton and AnotherSingleton are identical");



#========================================================================
#
# the two classes derived from UniqueSingleton should be mutually 
# exclusive.  That is, the instances of different classes should be 
# the same.
#
#========================================================================

my $s7 = UniqueSingletonOne->instance();

print "UniqueSingetonOne instance 1: ",
    defined($s7) ? ok($s7) : not_ok('<undef>');

# test method calls (one() should work, two() shouldn't)
eval {
    $s7->one("This is one");
};
print $@
    ? not_ok($@)
    : ok("UniqueSingletonOne is correctly blessed (1/2)");

eval {
    $s7->two("This is two");
};
print $@
    ? ok("UniqueSingletonOne is correctly blessed (2/2)")
    : not_ok("Called a non-existant method on UniqueSingletonOne");


my $s8 = UniqueSingletonTwo->instance();

print "UniqueSingetonTwo instance 1: ",
    defined($s8) ? ok($s8) : not_ok('<undef>');

# test method calls (one() shouldn't work, two() should)
eval {
    $s8->two("This is two");
};
print $@
    ? not_ok($@)
    : ok("UniqueSingletonTwo is correctly blessed (1/2)");

eval {
    $s7->one("This is two");
};
print $@
    ? ok("UniqueSingletonTwo is correctly blessed (2/2)")
    : not_ok("Called a non-existant method on UniqueSingletonTwo");


print $s7 == $s8 
    ? ok("UniqueSingleton derived instances are identical")
    : not_ok("UniqueSingleton derived instances instances are unique");


#========================================================================
#
# ok/not_ok subs
#
#========================================================================

sub ok     { 
    return join('', @_ ? (@_, "\n") : (), "ok ", ++$loaded, "\n");
}

sub not_ok { 
    return join('', @_ ? (@_, "\n") : (), "not ok ", ++$loaded, "\n");
}



lib/Class/Singleton.pm
#============================================================================
#
# Class::Singleton.pm
#
# Implementation of a "singleton" module which ensures that a class has
# only one instance and provides global access to it.  For a description 
# of the Singleton class, see "Design Patterns", Gamma et al, Addison-
# Wesley, 1995, ISBN 0-201-63361-2
#
# Written by Andy Wardley <abw@wardley.org>
#
# Copyright (C) 1998 Canon Research Centre Europe Ltd.
# Copyright (C) 1998-2008 Andy Wardley.  All rights reserved.
# Copyright (C) 2014 Steve Hay.  All rights reserved.
#
# This module is free software; you can redistribute it and/or modify it under
# the same terms as Perl itself, i.e. under the terms of either the GNU General
# Public License or the Artistic License, as specified in the F<LICENCE> file.
#
#============================================================================

package Class::Singleton;
use 5.008001;
use strict;
use warnings;

our $VERSION = 1.7;
my %_INSTANCES = ();


#========================================================================
#
# instance()
#
# Module constructor.  Creates an Class::Singleton (or derived) instance 
# if one doesn't already exist.  The instance reference is stored in the
# %_INSTANCES hash of the Class::Singleton package.  The impact of this is
# that you can create any number of classes derived from Class::Singleton
# and create a single instance of each one.  If the instance reference
# was stored in a scalar $_INSTANCE variable, you could only instantiate 
# *ONE* object of *ANY* class derived from Class::Singleton.  The first
# time the instance is created, the _new_instance() constructor is called 
# which simply returns a reference to a blessed hash.  This can be 
# overloaded for custom constructors.  Any additional parameters passed to 
# instance() are forwarded to _new_instance().
#
# Returns a reference to the existing, or a newly created Class::Singleton
# object.  If the _new_instance() method returns an undefined value
# then the constructer is deemed to have failed.
#
#========================================================================

sub instance {
    my $class = shift;
    
    # already got an object
    return $class if ref $class;

    # we store the instance against the $class key of %_INSTANCES
    my $instance = $_INSTANCES{$class};
    unless(defined $instance) {
        $_INSTANCES{$class} = $instance = $class->_new_instance(@_);
    }
    return $instance;
}


#=======================================================================
# has_instance()
#
# Public method to return the current instance if it exists.
#=======================================================================

sub has_instance {
    my $class = shift;
    $class = ref $class || $class;
    return $_INSTANCES{$class};
}


#========================================================================
# _new_instance(...)
#
# Simple constructor which returns a hash reference blessed into the 
# current class.  May be overloaded to create non-hash objects or 
# handle any specific initialisation required.
#========================================================================

sub _new_instance {
    my $class = shift;
    my %args  = @_ && ref $_[0] eq 'HASH' ? %{ $_[0] } : @_;
    bless { %args }, $class;
}


#========================================================================
# END()
#
# END block to explicitly destroy all Class::Singleton objects since
# destruction order at program exit is not predictable. See CPAN RT
# bugs #23568 and #68526 for examples of what can go wrong without this.
#========================================================================

END {
    # dereferences and causes orderly destruction of all instances
    undef(%_INSTANCES);
}


1;

__END__

=head1 NAME

Class::Singleton - Implementation of a "Singleton" class 

=head1 SYNOPSIS

    use Class::Singleton;
    
    my $one = Class::Singleton->instance();   # returns a new instance
    my $two = Class::Singleton->instance();   # returns same instance

=head1 DESCRIPTION

This is the C<Class::Singleton> module.  A Singleton describes an object class
that can have only one instance in any system.  An example of a Singleton
might be a print spooler or system registry.  This module implements a
Singleton class from which other classes can be derived.  By itself, the
C<Class::Singleton> module does very little other than manage the instantiation
of a single object.  In deriving a class from C<Class::Singleton>, your module 
will inherit the Singleton instantiation method and can implement whatever
specific functionality is required.

For a description and discussion of the Singleton class, see 
"Design Patterns", Gamma et al, Addison-Wesley, 1995, ISBN 0-201-63361-2.

=head2 Using the Class::Singleton Module

To import and use the C<Class::Singleton> module the following line should 
appear in your Perl program:

    use Class::Singleton;

The L<instance()> method is used to create a new C<Class::Singleton> instance,
or return a reference to an existing instance. Using this method, it is only
possible to have a single instance of the class in any system.

    my $highlander = Class::Singleton->instance();

Assuming that no C<Class::Singleton> object currently exists, this first call
to L<instance()> will create a new C<Class::Singleton> and return a reference
to it. Future invocations of L<instance()> will return the same reference.

    my $macleod    = Class::Singleton->instance();

In the above example, both C<$highlander> and C<$macleod> contain the same
reference to a C<Class::Singleton> instance.  There can be only one.

=head2 Deriving Singleton Classes

A module class may be derived from C<Class::Singleton> and will inherit the 
L<instance()> method that correctly instantiates only one object.

    package PrintSpooler;
    use base 'Class::Singleton';
    
    # derived class specific code
    sub submit_job {
        ...
    }
    
    sub cancel_job {
        ...
    }

The C<PrintSpooler> class defined above could be used as follows:

    use PrintSpooler;
    
    my $spooler = PrintSpooler->instance();
    
    $spooler->submit_job(...);

The L<instance()> method calls the L<_new_instance()> constructor method the
first and only time a new instance is created. All parameters passed to the
L<instance()> method are forwarded to L<_new_instance()>. In the base class
the L<_new_instance()> method returns a blessed reference to a hash array
containing any arguments passed as either a hash reference or list of named 
parameters. 

    package MyConfig;
    use base 'Class::Singleton';
    
    sub foo {
        shift->{ foo };
    }
    
    sub bar {
        shift->{ bar };
    }
    
    package main;
    
    # either: hash reference of named parameters
    my $config = MyConfig->instance({ foo => 10, bar => 20 });
    
    # or: list of named parameters
    my $config = MyConfig->instance( foo => 10, bar => 20 );
    
    print $config->foo();   # 10
    print $config->bar();   # 20

Derived classes may redefine the L<_new_instance()> method to provide more
specific object initialisation or change the underlying object type (to a list
reference, for example).

    package MyApp::Database;
    use base 'Class::Singleton';
    use DBI;
    
    # this only gets called the first time instance() is called
    sub _new_instance {
        my $class = shift;
        my $self  = bless { }, $class;
        my $db    = shift || "myappdb";    
        my $host  = shift || "localhost";
        
        $self->{ DB } = DBI->connect("DBI:mSQL:$db:$host")
            || die "Cannot connect to database: $DBI::errstr";
        
        # any other initialisation...
        
        return $self;
    }

The above example might be used as follows:

    use MyApp::Database;
    
    # first use - database gets initialised
    my $database = MyApp::Database->instance();

Some time later on in a module far, far away...

    package MyApp::FooBar
    use MyApp::Database;
    
    # this FooBar object needs access to the database; the Singleton
    # approach gives a nice wrapper around global variables.
    
    sub new {
        my $class = shift;
        bless {
            database => MyApp::Database->instance(),
        }, $class;
    }

The C<Class::Singleton> L<instance()> method uses a private hash to store
a reference to any existing instance of the object, keyed against the derived
class package name.

This allows different classes to be derived from C<Class::Singleton> that can
co-exist in the same system, while still allowing only one instance of any one
class to exist. For example, it would be possible to derive both
'C<PrintSpooler>' and 'C<MyApp::Database>' from C<Class::Singleton> and have a
single instance of I<each> in a system, rather than a single instance of
I<either>.

You can use the L<has_instance()> method to find out if a particular class 
already has an instance defined.  A reference to the instance is returned or
C<undef> if none is currently defined.

    my $instance = MyApp::Database->has_instance()
        || warn "No instance is defined yet";

=head2 Methods

=over 4

=item instance()

This method is called to return a current object instance or create a new
one by calling L<_new_instance()>.

=item has_instance()

This method returns a reference to any existing instance or C<undef> if none
is defined.

    my $testing = MySingleton1->has_instance()
        || warn "No instance defined for MySingleton1";

=item _new_instance()

This "private" method is called by L<instance()> to create a new object
instance if one doesn't already exist. It is not intended to be called
directly (although there's nothing to stop you from calling it if you're
really determined to do so).

It creates a blessed hash reference containing any arguments passed to the
method as either a hash reference or list of named parameters.

    # either: hash reference of named parameters
    my $example1 = MySingleton1->new({ pi => 3.14, e => 2.718 });

    # or: list of named parameters
    my $example2 = MySingleton2->new( pi => 3.14, e => 2.718 );

It is important to remember that the L<instance()> method will I<only> call
the I<_new_instance()> method once, so any arguments you pass may be silently
ignored if an instance already exists. You can use the L<has_instance()>
method to determine if an instance is already defined.

=back

=head1 EXPORTS

I<None>.

=head1 KNOWN BUGS

I<None>.

=head1 FEEDBACK

Patches, bug reports, suggestions or any other feedback is welcome.

Patches can be sent as GitHub pull requests at
L<https://github.com/steve-m-hay/Class-Singleton/pulls>.

Bug reports and suggestions can be made on the CPAN Request Tracker at
L<https://rt.cpan.org/Public/Bug/Report.html?Queue=Class-Singleton>.

Currently active requests on the CPAN Request Tracker can be viewed at
L<https://rt.cpan.org/Public/Dist/Display.html?Status=Active;Queue=Class-Singleton>.

Please test this distribution.  See CPAN Testers Reports at
L<https://www.cpantesters.org/> for details of how to get involved.

Previous test results on CPAN Testers Reports can be viewed at
L<https://www.cpantesters.org/distro/C/Class-Singleton.html>.

Please rate this distribution on CPAN Ratings at
L<https://cpanratings.perl.org/rate/?distribution=Class-Singleton>.

=head1 AVAILABILITY

The latest version of this module is available from CPAN (see
L<perlmodlib/"CPAN"> for details) at

L<https://metacpan.org/release/Class-Singleton> or

L<https://www.cpan.org/authors/id/S/SH/SHAY/> or

L<https://www.cpan.org/modules/by-module/Class/>.

The latest source code is available from GitHub at
L<https://github.com/steve-m-hay/Class-Singleton>.

=head1 INSTALLATION

See the F<INSTALL> file.

=head1 AUTHOR

Andy Wardley E<lt>L<abw@wardley.org|mailto:abw@wardley.org>E<gt>
L<http://wardley.org/>.

Thanks to Andreas Koenig for providing some significant speedup patches and
other ideas.

Steve Hay E<lt>L<shay@cpan.org|mailto:shay@cpan.org>E<gt> is now maintaining
Class::Singleton as of version 1.5.

=head1 COPYRIGHT

Copyright (C) 1998 Canon Research Centre Europe Ltd.

Copyright (C) 1998-2008 Andy Wardley.  All rights reserved.

Copyright (C) 2014, 2020 Steve Hay.  All rights reserved.

=head1 LICENCE

This module is free software; you can redistribute it and/or modify it under the
same terms as Perl itself, i.e. under the terms of either the GNU General Public
License or the Artistic License, as specified in the F<LICENCE> file.

=head1 VERSION

Version 1.7

=head1 DATE

TODO

=head1 HISTORY

See the F<Changes> file.

=cut


Makefile.PL
#!perl
#===============================================================================
#
# Makefile.PL
#
# DESCRIPTION
#   Makefile creation script.
#
# COPYRIGHT
#   Copyright (C) 2014-2015 Steve Hay.  All rights reserved.
#
# LICENCE
#   This script is free software; you can redistribute it and/or modify it under
#   the same terms as Perl itself, i.e. under the terms of either the GNU
#   General Public License or the Artistic License, as specified in the LICENCE
#   file.
#
#===============================================================================

use 5.008001;

use strict;
use warnings;

use ExtUtils::MakeMaker 6.64;
use ExtUtils::MakeMaker qw(WriteMakefile);

#===============================================================================
# MAIN PROGRAM
#===============================================================================

MAIN: {
    WriteMakefile(
        NAME          => 'Class::Singleton',
        ABSTRACT_FROM => 'lib/Class/Singleton.pm',
        AUTHOR        => 'Andy Wardley <abw@wardley.org>, Steve Hay <shay@cpan.org>',
        LICENSE       => 'perl_5',
        VERSION_FROM  => 'lib/Class/Singleton.pm',

        META_MERGE => {
            'meta-spec' => {
                version => 2
            },

            resources => {
                repository => {
                    type => 'git',
                    url  => 'https://github.com/steve-m-hay/Class-Singleton.git'
                }
            },

            optional_features => {
                changestest => {
                    description => 'Changes testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::CPAN::Changes' => '0'
                            }
                        }
                    }
                },

                critictest => {
                    description => 'Perl::Critic testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Perl::Critic' => '0'
                            }
                        }
                    }
                },

                podtest => {
                    description => 'POD testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Pod' => '1.00'
                            }
                        }
                    }
                },

                podcoveragetest => {
                    description => 'POD coverage testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Pod::Coverage' => '0.08'
                            }
                        }
                    }
                }
            }
        },

        MIN_PERL_VERSION => '5.008001',

        CONFIGURE_REQUIRES => {
            'ExtUtils::MakeMaker' => '6.64',
            'perl'                => '5.008001',
            'strict'              => '0',
            'warnings'            => '0'
        },

        TEST_REQUIRES => {
            'Test::More' => '0',
            'base'       => '0'
        },

        PREREQ_PM => {
            'strict'   => '0',
            'warnings' => '0'
        },

        dist => {
            PREOP   => 'find $(DISTVNAME) -type d -print|xargs chmod 0755 && ' .
                       'find $(DISTVNAME) -type f -print|xargs chmod 0644',
            TO_UNIX => 'find $(DISTVNAME) -type f -print|xargs dos2unix'
        }
    );
}

#===============================================================================
